### PR TITLE
[RFC] Define a Phase2_tracker_postTDR era for development purposes

### PR DESCRIPTION
**Pull request by emiglior**
In the context of the Tracker studies for phase2, we are investigating configurations of the digitizer(s) and of the local reco with parameters reflecting the evolution in the design of the detector hardware post-TDR.
The values of these parameters are not final yet but nevertheless we would like to have them in release as we would need to validate them on RelVal samples.
In order to keep in the same CMSSW release the TDR and the post-TDR settings, I would propose to introduce a "Tracker post-TDR" era. 
This PR enables to activate this post-TDR era from the cmsDriver command line option
'--era Phase2_tracker_postTDR' 
but I have not created the corresponding workflows yet.
Thanks in advance for the feedback on the strategy and on the way I have implemented.
Ernesto 

@kpedro88 @slava77: this it not for integration, only for discussion.
@pmaksim1 @boudoul FYI

### CODE REVIEW

**Comment by kpedro88** [comment]
This is acceptable in principle - similar to some of the special Run2 tracking eras.

The special workflows for the new era should be implemented in the same way:
1. define the applicable steps, offset, and workflow name suffix in the `upgradeSteps` dictionary here: https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py
2. make the step variations here (i.e. change to use the new era): https://github.com/cms-sw/cmssw/blob/78e25c56c86bae6d05ba616e797d5ec67a13e321/Configuration/PyReleaseValidation/python/relval_steps.py#L2568
3. create the workflow for just one sample, e.g. 2023 TTbar_14TeV as shown here: https://github.com/cms-sw/cmssw/blob/78e25c56c86bae6d05ba616e797d5ec67a13e321/Configuration/PyReleaseValidation/python/relval_upgrade.py#L65
4. forward the new workflow numbers into the regular matrix if they are intended to run frequently (e.g. in IBs): https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2023.py

I admit this is a somewhat cumbersome procedure, but it's better than it used to be.

File: SimGeneral/MixingModule/python/pixelDigitizer_cfi.py [hunk 1]
```diff
@@ -22,6 +22,15 @@
...
 
 from SimTracker.SiPhase2Digitizer.phase2TrackerDigitizer_cfi import phase2TrackerDigitizer as _phase2TrackerDigitizer
+
 from Configuration.Eras.Modifier_phase2_tracker_cff import phase2_tracker
 phase2_tracker.toReplaceWith(pixelDigitizer, _phase2TrackerDigitizer)
 
+# tracker phase2 post-TDR
+from Configuration.StandardSequences.Eras import eras
+eras.phase2_tracker_postTDR.toModify(_phase2TrackerDigitizer, PixelDigitizerAlgorithm  = dict(TanLorentzAnglePerTesla_Barrel = 0.0459))
+eras.phase2_tracker_postTDR.toModify(_phase2TrackerDigitizer, PixelDigitizerAlgorithm  = dict(TanLorentzAnglePerTesla_Endcap = 0.0459))
```
Comment (kpedro88) [inline SimGeneral/MixingModule/python/pixelDigitizer_cfi.py]: the change in this line can be combined with the previous line (multiple items can be specified within `dict()`).

Comment (makortel) [reply SimGeneral/MixingModule/python/pixelDigitizer_cfi.py]: Assuming `phase2_tracker_postTDR` will always require `phase2_tracker`, these customizations should be moved to 
https://github.com/cms-sw/cmssw/blob/78e25c56c86bae6d05ba616e797d5ec67a13e321/SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py#L174
(before the premixing customization). Then this file would not need any changes (`phase2_tracker` already replaces `pixelDigitizer` with `_phase2TrackerDigitizer`), and the customizations will work automatically correctly with the upcoming premixing.

**Comment by emiglior** [comment]
@kpedro88 @makortel please check if your suggestions were implemented properly
New workflows not implemented yet

**Comment by kpedro88** [comment]
looks okay to me so far

**Review by makortel — COMMENTED** [review]
Functionally looks ok now, thanks. I have still two, mostly stylistic comments.

File: RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py [hunk 2]
```diff
@@ -72,3 +72,16 @@
...
+from Configuration.Eras.Modifier_phase2_tracker_postTDR_cff import phase2_tracker_postTDR
+phase2_tracker_postTDR.toModify(PixelCPEGenericESProducer, 
+  useLAWidthFromDB = False,
+  UseErrorsFromTemplates = False,
+  LoadTemplatesFromDB = False,
+  TruncatePixelCharge = False,
+  IrradiationBiasCorrection = False,
+  DoCosmics = False,
+  Upgrade = cms.bool(True),
+  lAOffset = cms.double(0.0459)
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py]: only this ```lAOffset = 0.0459``` is needed, everything else is unnecessary copy-paste of the phase2_tracker mods

**Review by slava77 — COMMENTED** [review]
I see that only PixelCPEGenericESProducer.lAOffset and phase2TrackerDigitizer.PixelDigitizerAlgorithm.TanLorentzAnglePerTesla_{Barrel,Endcap} are actually modified.

A new era for this in a global space seems like an overkill, it's not really a setup for a new detector.
The solution for simple switches of configuration is to simply add the process modifier to the cmsDriver on top of the standard era.
Also, the preferred (but not really widely used yet) solution is to add such technical/test modifiers to Configuration/ProcessModifiers/python

I don't quite understand why we can't simply update the configuration in the default phase2_tracker.

File: Configuration/Eras/python/Era_Phase2_tracker_postTDR_cff.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import FWCore.ParameterSet.Config as cms
+
+from Configuration.Eras.Era_Phase2_cff import Phase2
+from Configuration.Eras.Modifier_phase2_tracker_postTDR_cff import phase2_tracker_postTDR
+
+Phase2_tracker_postTDR = cms.ModifierChain(Phase2, phase2_tracker_postTDR)
```
[the remark this answers is not in the document]
Comment (kpedro88) [reply Configuration/Eras/python/Era_Phase2_tracker_postTDR_cff.py]: this follows the current (admittedly informal) standard: top-level Eras are capitalized, individual modifiers are lowercase.

**Comment by slava77** [comment]
>  In order to keep in the same CMSSW release the TDR and the post-TDR settings, I would propose to introduce a "Tracker post-TDR" era.

What is the default? I assume the default is the post-TDR.
If for some ephemeral reason the TDR parameters should be maintained, it makes more sense to me to define a special "TDR" phase2_tracker_TDR modifier.
The simplest maintenance of this would be to assume that this phase2_tracker_TDR is applied on top of the phase2_tracker defaults.

IIUC, in a long run we can not assume that the full TDR will have much meaning in a release other than where the TDR samples were made. 
So, the question is: for how long are you planning to keep it alive and updated in the master branch?

File: Configuration/Eras/python/Era_Phase2_tracker_postTDR_cff.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (slava77) [reply Configuration/Eras/python/Era_Phase2_tracker_postTDR_cff.py]: If I'm not mistaken the only cases of the era = capitalize(modifier) is Phase2_timing and Phase2_timing_layer. More of an exception than a standard.

[the remark this answers is not in the document]
Comment (kpedro88) [reply Configuration/Eras/python/Era_Phase2_tracker_postTDR_cff.py]: Ah, that's fair - I was focusing on the capitalization standard, didn't quite notice that they're exactly the same words modulo the capitalization.

**Comment by pmaksim1** [comment]
@slava77

Right now, you are right that only the Lorentz angle (LA) in the pixel
Phase 2 digitizer and in Pixel CPE Generic is modified.  However, soon
there will be more:

- CPE Generic will load GenError DB object corresponding to the same LA
- CPE TemplateReco will be turned on and load the templates corresponding
to the same LA (they go together)
- we'll then want to be more realistic and also load the 2D templates (both
data and MC) into the Phase 2 digitizer, also corresponding to that same LA
- finally, Pixel FastSim resolution histograms will also need to match all
this LA as well.

So this may affect multiple packages and several config files, and require
corresponding GT and such, etc.

**Comment by slava77** [comment]
On 4/27/18 9:57 PM, Petar Maksimovic wrote:
> @slava77
> 
> Right now, you are right that only the Lorentz angle (LA) in the pixel
> Phase 2 digitizer and in Pixel CPE Generic is modified. However, soon
> there will be more:
> 
> - CPE Generic will load GenError DB object corresponding to the same LA
> - CPE TemplateReco will be turned on and load the templates corresponding
> to the same LA (they go together)
> - we'll then want to be more realistic and also load the 2D templates (both
> data and MC) into the Phase 2 digitizer, also corresponding to that same LA
> - finally, Pixel FastSim resolution histograms will also need to match all
> this LA as well.
> 
> So this may affect multiple packages and several config files, and require
> corresponding GT and such, etc.
> 
> 

OK, makes sense.
Clearly, the plans are to keep making progress after the TDR is out.
It also sounds like these plans imply that the progress made will (at
least eventually) apply to the standard setup.
So, making all these changes to the default phase2_tracker makes the
most sense to me.

File: SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py [hunk 4]
```diff
@@ -172,6 +172,15 @@
...
 )
 
+# tracker phase2 post-TDR
+from Configuration.StandardSequences.Eras import eras
+eras.phase2_tracker_postTDR.toModify(phase2TrackerDigitizer, 
+                                     PixelDigitizerAlgorithm  = dict(
+        TanLorentzAnglePerTesla_Barrel = 0.0459,
+        TanLorentzAnglePerTesla_Endcap = 0.0459
+        )
+)
```
[the remark this answers is not in the document]
Comment (emiglior) [reply SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py]: @makortel could you point me to some reference for the indentation?

File: RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py (hunk 2 above)
Comment (emiglior) [reply RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py]: @slava77 I am bit confused: should I define the phase2_tracker_postTDR mod as a clone of phase2_tracker mod and then add lAOffset?

File: SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (perrotta) [reply SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py]: @emiglior : e.g. the few lines rigth below these added ones

File: RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py (hunk 2 above)
Comment (slava77) [reply RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py]: your current implementation is for phase2_tracker_postTDR to add changes on top of phase2_tracker.

**Comment by slava77** [comment]
@emiglior 
please clarify if some more activity is expected with this "RFC" PR.

**Comment by emiglior** [comment]
> @emiglior
> please clarify if some more activity is expected with this "RFC" PR.

@slava77 I am planning to look at your suggestions but only at the end of this week

About the need for a "Tracker post-TDR" era (or whatever it will be labeled), as Tracker we really would like to have in the same CMSSW release the TDR version "T5" (frozen from geometry to local reco) and a development version.
There are essentially two reasons for this request:
1. Many of the hardware options we plan to describe in the development version are not yet final and we need results from full-sim RelVal to guide the choice for the final design. Having both versions in the same release would make the comparison easier.
2. The frozen TDR version still describes the state-of-the-art of the phase2 Tracker. Therefore in case a description of a realistic phase2 Tracker is required for some MC production, the TDR version will be at any time a good choice ready to be deployed

**Comment by slava77** [comment]
@emiglior 
In case you have collected all comments (as a part of RFC), perhaps it's time to turn this into a normal PR or close this.

**Comment by slava77** [comment]
@emiglior
In case you have collected all comments (as a part of RFC), perhaps it's time to turn this into a normal PR or close this.

**Comment by emiglior** [comment]
Thanks for all the inputs received.
My understanding is that the proposed changes are not mature yet to become a PR now but unfortunately I cannot work on it in on a short time scale.

**Pull request closed without merging**